### PR TITLE
Hash compare by identity

### PR DESCRIPTION
**Pull request by asterite**
Implements #8329

Has a commit from #8450 but I can later rebase on top of `master`, or just rewrite git history.

### CODE REVIEW

File: src/hash.cr [hunk 1]
```diff
@@ -893,17 +896,73 @@ class Hash(K, V)
...
+            K < Enum %}
+      entry.key == key && entry.hash != 0_u32
+    {% else %}
+      entry.hash == hash && entry_matches?(entry, key)
+    {% end %}
+  end
+
+  private def entry_matches?(entry, key)
+    entry_key = entry.key
+    if @compare_by_identity && entry_key.responds_to?(:object_id)
```
Comment (straight-shoota) [inline src/hash.cr]: What if `@compare_by_identity` is true and `entry` responds to `object_id` but `entry_key` doesn't? Shouldn't that also return false even if `entry_key == key`?

Comment (asterite) [reply src/hash.cr]: I don't know. If `entry_key` is an int you still want ints to compare the same way.

I also thought about raising a compile-time error when using `compare_by_identity` on a hash where keys don't respond to `object_id`, but it involved a chunk of macro code. Would something like that be better?

In general, I wouldn't worry too much about this. If you use `compare_by_identity` you generally know what you are doing and you'll be already working with reference types, or reference-like types. But we can try adding the macro check.

Comment (straight-shoota) [reply src/hash.cr]: I'm not sure what's the best semantics here. But I'd try to be consistent.

Currently, when comparing by identity the behaviour for differing answers to `responds_to?(:object_id)` is inconsistent:
* when `entry_key.responds_to?(:object_id) && !key.responds_to?(:object_id)` the result is `false`
* when  `!entry_key.responds_to?(:object_id) && key.responds_to?(:object_id)` the result is `entry_key == key`

Comment (straight-shoota) [reply src/hash.cr]: I'm not sure what's the best semantics here. But I'd try to be consistent.

Currently, when comparing by identity the behaviour for differing answers to `responds_to?(:object_id)` is inconsistent:
* when `entry_key.responds_to?(:object_id) && !key.responds_to?(:object_id)` the result is `false`
* when  `!entry_key.responds_to?(:object_id) && key.responds_to?(:object_id)` the result is `entry_key == key`

Comment (asterite) [reply src/hash.cr]: Sounds good, I'll at least make it consistent

Comment (asterite) [reply src/hash.cr]: Done! Let me know what you think.

**Comment by asterite** [comment]
> Shouldn't dup and clone preserve the compare_by_identity?

Great catch! Done

**Comment by bew** [comment]
Is it intentional to not have a way to switch back to "compare by hash" ?

**Comment by straight-shoota** [comment]
Switching back an forth on the same instance would be confusing. `compare_by_identity` should be configured only once and prior to any usage of the hash. Ideally, that's in the constructor and I think we should actually make it a constructor argument (see discussion in #8329).

**Pull request merged**